File: exe/covert_search_dp/covert_search_dp.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * covert_search_dp.c
 *
 * Code generation for function 'covert_search_dp'
 *
 */

/* Include files */
#include "covert_search_dp.h"
#include "covert_search_dp_data.h"
#include "covert_search_dp_emxutil.h"
#include "covert_search_dp_initialize.h"
#include "covert_search_dp_types.h"
#include "fileManager.h"
#include "fread.h"
#include "rand.h"
#include "randn.h"
#include "randsample.h"
#include "tic.h"
#include "toc.h"
#include <math.h>

/* Function Definitions */
void covert_search_dp(double trials, const char dpmapfp[9], const char priorhfp
                      [10])
{
  static double unusedExpr[360000];
  emxArray_int8_T *tpaMat;
  emxArray_real_T c_priorh;
  emxArray_real_T *b_fileid;
  emxArray_real_T *b_priorh;
  emxArray_real_T *b_tpaMat;
  emxArray_real_T *fileid;
  emxArray_real_T *priorh;
  double x_tmp;
  int d_priorh[1];
  int i;
  int mti;
  unsigned int r;
  signed char c_fileid;
  if (!isInitialized_covert_search_dp) {
    covert_search_dp_initialize();
  }

  /* codegen */
  /*  COVERT_SEARCH_DP runs the search model with specified parameters and */
  /*  number of trials. */
  /*  */
  /*  COVERT_SEARCH_DP calls the following routines: */
  /*  mk_prior   make the prior prob distribution of target location */
  /*  mk_amap    make a map of target amplitude (may change if contrast fixed) */
  /*  mk_fov     make the scalar foveation map */
  /*  mk_tarloc  randomly sample target location from prior */
  /*  mk_lr      compute log likelihood ratio */
  /*  */
  /* note that (i,j) = (y,x) */
  /*  note:  xcorr2 = conv2(a, rot90(conj(b),2));  can add 'same' to this */
  tic();
  r = 1U;
  state[0] = 1U;
  for (mti = 0; mti < 623; mti++) {
    r = ((r ^ r >> 30U) * 1812433253U + mti) + 1U;
    state[mti + 1] = r;
  }

  emxInit_real_T(&priorh, 2);
  emxInit_int8_T(&tpaMat, 1);
  emxInit_real_T(&fileid, 2);
  emxInit_real_T(&b_fileid, 2);
  state[624] = 624U;
  c_fileid = cfopen(dpmapfp, "rb");
  b_fread(c_fileid, fileid);
  c_fileid = b_cfopen(priorhfp, "rb");
  b_fread(c_fileid, priorh);
  c_fileid = c_cfopen("rtmp.bin", "rb");
  c_fread(c_fileid, b_fileid);

  /*  Setup some generic parameters */
  /*  Fixed parameters */
  /*  */
  /*  Create effective prior */
  x_tmp = floor(trials / 2.0);
  i = tpaMat->size[0];
  tpaMat->size[0] = (int)x_tmp + (int)x_tmp;
  emxEnsureCapacity_int8_T(tpaMat, i);
  mti = (int)x_tmp;
  emxFree_real_T(&b_fileid);
  emxFree_real_T(&fileid);
  for (i = 0; i < mti; i++) {
    tpaMat->data[i] = 0;
  }

  mti = (int)x_tmp;
  for (i = 0; i < mti; i++) {
    tpaMat->data[i + (int)x_tmp] = 1;
  }

  emxInit_real_T(&b_tpaMat, 2);
  emxInit_real_T(&b_priorh, 1);

  /*  1 for target present, 0 for absent. */
  b_rand(tpaMat->size[0], b_tpaMat);

  /*  permute isn't necessary becuase this search with no memory. we permute anyway. */
  /*  integer pixel locations */
  mti = priorh->size[0] * priorh->size[1];
  c_priorh = *priorh;
  d_priorh[0] = mti;
  c_priorh.size = &d_priorh[0];
  c_priorh.numDimensions = 1;
  randsample(priorh->size[0] * priorh->size[1], trials, &c_priorh, b_priorh);

  /*  Run simulation loop */
  i = (int)trials;
  emxFree_real_T(&b_priorh);
  emxFree_real_T(&b_tpaMat);
  emxFree_int8_T(&tpaMat);
  emxFree_real_T(&priorh);
  for (mti = 0; mti < i; mti++) {
    /*  background number */
    randn(unusedExpr);

    /*  human experiments have 30 independent pix per degree. upsample via nearest neighbor to run simulations in 120 pix per degree. */
    /*  sneaky way to increase the matrix with blocks -- Kronecker product */
    /*  compute normalized responses */
    /*  compute log likelihood ratio of target present, */
    /*  location of max normalized response, and max normalized response */
    /* log(sum(eslr(:), 'omitnan')); */
    /*  record data     */
  }

  /*  Structured Data Storage */
  /* results = 1; */
  /* results = table(trial, stimY, stimX, tPresent, ymax, xmax, llr, slpmx,... */
  /*     'VariableNames', {'trial', 'stimY', 'stimX', 'tPresent', 'ymax', 'xmax', 'llr', 'slpmx'}); */
  toc();
}

/* End of code generation (covert_search_dp.c) */

File: mat/covert_search_dp.m
function covert_search_dp(trials, dpmapfp, priorhfp, seed_num) %codegen
% COVERT_SEARCH_DP runs the search model with specified parameters and
% number of trials.
%
% COVERT_SEARCH_DP calls the following routines:
% mk_prior   make the prior prob distribution of target location
% mk_amap    make a map of target amplitude (may change if contrast fixed)
% mk_fov     make the scalar foveation map
% mk_tarloc  randomly sample target location from prior
% mk_lr      compute log likelihood ratio
%
%note that (i,j) = (y,x)
% note:  xcorr2 = conv2(a, rot90(conj(b),2));  can add 'same' to this

tic;
rng(seed_num)
% preliminaries
dataTotal = zeros(trials, 7);
expandMat = ones(4,4);

%% load data
fileID = fopen(dpmapfp);
dpmap  = fread(fileID, [2400, 2400], 'double');
dpmap  = imrotate(dpmap, 180);

fileID = fopen(priorhfp);
priorh = fread(fileID, [2400, 2400], 'double');

fileID = fopen('rtmp.bin');
rtmp   = fread(fileID, [135, 135], 'double');

%% Setup some generic parameters for size
sz      = floor(600);
szrtmp = size(rtmp); 
rt     = floor(szrtmp(1)/2); 

%%
lnepriorh       = log(priorh); % Create effective prior
half_dpmap      = 0.5 .* dpmap.^2;
half_dpmap_with_eprior = half_dpmap - lnepriorh;



tpaMat          = [zeros(floor(trials/2),1); ones(floor(trials/2),1)]; % 1 for target present, 0 for absent.
tpaMat          = tpaMat(randperm(length(tpaMat))); % permute isn't necessary becuase this search with no memory. we permute anyway.

N               = length(priorh(:)); % integer pixel locations
k               = randsample(N,trials,true,priorh(:));
[pre_yt,pre_xt] = ind2sub(size(priorh), k);


%% Run simulation loop
for k = 1:trials            % background number
    re  = randn(sz, sz); % human experiments have 30 independent pix per degree. upsample via nearest neighbor to run simulations in 120 pix per degree.
    re  = kron(re, expandMat); % sneaky way to increase the matrix with blocks -- Kronecker product
    tpa = tpaMat(k);
    if tpa           % if target present add target responses
        xt = pre_xt(k);
        yt = pre_yt(k);
        re(yt-rt:yt+rt,xt-rt:xt+rt) = ...
            re(yt-rt:yt+rt,xt-rt:xt+rt) + abs(dpmap(yt-rt:yt+rt,xt-rt:xt+rt).*rtmp); % create the local target response
    else
        tpa = 0;
        xt  = 0;
        yt  = 0;
    end    
    
    % compute normalized responses
    % compute log likelihood ratio of target present,
    % location of max normalized response, and max normalized response
    norm_resp = re .* dpmap - half_dpmap_with_eprior;
    
    llr    = -1;%log(sum(eslr(:), 'omitnan'));
    slpmx  = max(norm_resp(:));
    [Y, X] = find(norm_resp== max(norm_resp(:))); % since we use the max rule, find the max.

    % record data    
    dataTotal(k, :) = [yt, xt, tpa, Y, X, llr, slpmx];
end
data = dataTotal;

%% Structured Data Storage
stimY = data(:, 1);
stimX = data(:, 2);
tPresent = data(:, 3);
ymax     = data(:, 4);
xmax     = data(:, 5);
llr      = data(:, 6);
slpmx    = data(:, 7);
trial = (1:length(slpmx))';

%results = 1;
%results = table(trial, stimY, stimX, tPresent, ymax, xmax, llr, slpmx,...
%    'VariableNames', {'trial', 'stimY', 'stimX', 'tPresent', 'ymax', 'xmax', 'llr', 'slpmx'});
toc;